src/main/java/duke/logic/TaskList.java

```diff
@@ -0,0 +1,196 @@
+package duke.logic;
+
+import duke.task.*;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Scanner;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * TaskList stores user tasks and manages them.
+ *
+ * @author totsukatomofumi
+ */
+public class TaskList extends ArrayList<Task> {
+    /** Storage object for writing task history to a file. */
+    private Storage storage;
+
+    /**
+     * HashMap to contain set chars representing task types and function to return a new task of that
+     * corresponding type pairs.
+     */
+    private static final HashMap<Character,
+            BiFunction<Integer, Integer, Function<String, Task>>> taskMap = new HashMap<>();
+
+    static {
+        TaskList.taskMap.put('T', (index, length) -> line -> new ToDo(line.substring(index)));
+        TaskList.taskMap.put('D', (index, length) -> line -> new Deadline(line.substring(index, index + length),
+                        LocalDate.parse(line.substring(index + length))));
+        TaskList.taskMap.put('E', (index, length) -> line -> new Event(line.substring(index, index + length),
+                LocalDate.parse(line.substring(index + length))));
+    }
+
+    private TaskList() {
+        super();
+    }
+
+    /**
+     * Constructs a task list.
+     *
+     * @param storage the storage object to be tied to this task list.
+     */
+    public TaskList(Storage storage) {
+        super();
+        this.storage = storage;
+        //retrieve or else clear the file
+        try {
+            this.retrieve();
+        } catch (IOException e) {   //thrown
+            //make sure file is not deleted, else make again
+            this.storage.createRequiredFiles();
+            this.storage.clear();
+        }
+    }
+
+    /**
+     * Checks if a task exists at that zero-based index.
+     * @param query the zero-based index.
+     * @return true if a task exists, else false.
+     */
+    public boolean exists(int query) {
+        return query < super.size() && query >= 0;
+    }
+    
+    private void updateStorage() {
+        if (this.storage != null) {
+            this.storage.update(this);
+        }
+    }
+
+    /**
+     * Removes the task at the specified position from this task list.
+     * @param index the index of the task to be removed.
+     * @return the task that was removed from the task list.
+     */
+    @Override
+    public Task remove(int index) {
+        Task temp = super.remove(index);
+        this.updateStorage();
+        return temp;
+    }
+
+    /**
+     * Appends the specified task to the end of the task list.
+     *
+     * @param task task to be appended to this task list.
+     * @return true.
+     */
+    @Override
+    public boolean add(Task task) {
+        boolean bool = super.add(task);
```

Comment:
Should variables be named this way? Consider a more descriptive name like taskAdded